### ISSUE TITLE
No modified.iso file created

### ISSUE THREAD

**Issue by rakeshtembhurne** [sundarnagarajan/rdp-thinbook-linux#43]
I created `ISO` folder and `in` and `out` folders in it. Copied the bash file, made it executable and ran it. Everything seems to be running well, except the output file is not created. The `out` folder is just empty. Tried it on **Ubuntu 18.04.3 LTS**.

```
# ./make_rdp_iso.sh 
All required packages are already installed
Required packages:
    grub-efi-ia32-bin grub-efi-amd64-bin grub-pc-bin grub2-common
    grub-common util-linux parted gdisk mount xorriso genisoimage
    squashfs-tools rsync git build-essential kernel-package fakeroot
    libncurses5-dev libssl-dev ccache libfile-fcntllock-perl curl

Required space :        10000000000
Available space:        20888199168

Pulling latest changes to bootutils
Already up to date.
Pulling latest changes to kernel_build
Already up to date.
Pulling latest changes to rdp-thinbook-linux
Already up to date.
Directory not found: /usr/local/bin/linuxutils
(extract_iso): Extracting ISO ... source.iso
(extract_iso): Completed
(extract_squashfs): Extracting squashfs <-- filesystem.squashfs
(extract_squashfs): Completed
toplevel.sh: No commands found: [iso_pre]
00_setup_locale.sh [chroot]: Starting
    Setting up locales
00_setup_locale.sh [chroot]: Completed
01_rebrand.sh [chroot]: Starting
    Updated /etc/issue
    Updated /etc/issue.net
    Updated /etc/lsb-release
    Updated /etc/os-release
    Updated /etc/dpkg/origins/default
01_rebrand.sh [chroot]: Completed
10_update_pci_usb_ids.sh [chroot]: Starting
    Updating PCI IDs
    Updating USB IDs
10_update_pci_usb_ids.sh [chroot]: Completed
11_copy_scripts.sh [chroot]: Starting
11_copy_scripts.sh [chroot]: Completed
12_copy_linuxutils.sh [chroot]: Starting
    LINUXUTILS_DIR not a directory: /remaster_tmp/commands/../linuxutils
12_copy_linuxutils.sh [chroot]: Completed
13_install_grub_packages.sh [chroot]: Starting
    Installing grub-efi-ia32-bin grub-efi-amd64-bin grub-pc-bin
    ii grub-efi-amd64-bin
    ii grub-efi-ia32-bin
    ii grub-pc-bin
13_install_grub_packages.sh [chroot]: Completed
20_install_kernels.sh [chroot]: Starting
    No deb files in /remaster_tmp/kernel-debs
20_install_kernels.sh [chroot]: Completed
21_install_kernel_ppa.sh [chroot]: Starting
    Executing: /tmp/apt-key-gpghome.BFG1wGq0tH/gpg.1.sh --recv-keys --keyserver keyserver.ubuntu.com ABF7C302A5B662BDE68E0EFE883F04480A577E61
    Installing cherrytux-image cherrytux-headers
    update-initramfs: Generating /boot/initrd.img-5.5.5
    Installed following packages:
        cherrytux-image
        cherrytux-headers
        linux-image-5.5.5
        linux-headers-5.5.5
21_install_kernel_ppa.sh [chroot]: Completed
22_remove_old_kernels.sh [chroot]: Starting
    Removing following packages:
        linux-image-5.3.0-28-generic
        linux-image-generic-hwe-18.04
        linux-modules-5.3.0-28-generic
        linux-modules-extra-5.3.0-28-generic
        linux-headers-5.3.0-28
        linux-headers-5.3.0-28-generic
        linux-headers-generic-hwe-18.04
        linux-signed-image-generic
        linux-generic
    No deb files in /remaster_tmp/kernel-debs
    Kernel-related packages remaining:
        linux-headers-5.5.5
        linux-image-5.5.5
22_remove_old_kernels.sh [chroot]: Completed
23_install_firmware.sh [chroot]: Starting
    './iwlwifi-Qu-c0-hr-b0-50.ucode' -> '/lib/firmware/././iwlwifi-Qu-c0-hr-b0-50.ucode'
    './iwlwifi-Qu-b0-hr-b0-50.ucode' -> '/lib/firmware/././iwlwifi-Qu-b0-hr-b0-50.ucode'
    './iwlwifi-QuZ-a0-hr-b0-50.ucode' -> '/lib/firmware/././iwlwifi-QuZ-a0-hr-b0-50.ucode'
    './iwlwifi-QuZ-a0-jf-b0-50.ucode' -> '/lib/firmware/././iwlwifi-QuZ-a0-jf-b0-50.ucode'
    './iwlwifi-Qu-b0-jf-b0-50.ucode' -> '/lib/firmware/././iwlwifi-Qu-b0-jf-b0-50.ucode'
    './iwlwifi-Qu-c0-jf-b0-50.ucode' -> '/lib/firmware/././iwlwifi-Qu-c0-jf-b0-50.ucode'
    './iwlwifi-cc-a0-50.ucode' -> '/lib/firmware/././iwlwifi-cc-a0-50.ucode'
23_install_firmware.sh [chroot]: Completed
31_install_sound.sh [chroot]: Starting
    
    UCM/bytcht-es8316 from https://github.com/kernins/linux-chwhi12/tree/master/configs/audio/ucm
    Remaining files under UCM are from https://github.com/plbossart/UCM
    
    If the original author has specified a license, these files are licensed
    under the respective licenses from the original author.
    
    If there is no LICENSE specified by the original author, these files are
    licensed under the GNU General Public License version 2.
    
    Blacklisting snd_hdmi_lpe_audio
31_install_sound.sh [chroot]: Completed
32_install_r8723_bluetooth.sh [chroot]: Ignoring non-executable file
33_set_wifi_reg_domain.sh [chroot]: Starting
    Setting DEFAULT Wifi regulatory domain to US
    Change /etc/default/crda to set REGDOMAIN to 2-letter ISO country code
    where you are operating Wifi to comply with local laws
    See: https://wireless.wiki.kernel.org/en/developers/Regulatory/CRDA
33_set_wifi_reg_domain.sh [chroot]: Completed
34_rtl8732bs_bt_install_firmware.sh [chroot]: Starting
    
    ---------------------------------------------------------------------------
    Making r8723bs Bluetooth work
    
    rtlbt_fw and rtlbt_config are from:
    https://github.com/lwfinger/rtl8723bs_bt.git
    This source is licensed under the same terms as the original.
    If there is no LICENSE specified by the original author, this
    source is hereby licensed under the GNU General Public License version 2.
    
    For license detils, see /root/hardware/LICENSE and 
    /root/hardware/bluetooth/rtl8723bs_bt/LICENSE
    ---------------------------------------------------------------------------
    
34_rtl8732bs_bt_install_firmware.sh [chroot]: Completed
42_update_all_packages.sh [chroot]: Ignoring non-executable file
51_apt_cleanup.sh [chroot]: Starting
51_apt_cleanup.sh [chroot]: Completed
01_update_iso_kernel.sh [iso_post]: Starting
    Using vmlinuz: vmlinuz-5.5.5
    Using initrd: initrd.img-5.5.5
    replaced vmlinuz
    replaced initrd
01_update_iso_kernel.sh [iso_post]: Completed
02_remove_existing_grub_efi.sh [iso_post]: Starting
    Removing old image file: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/efi.img
    Removing grub module dir: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/x86_64-efi
02_remove_existing_grub_efi.sh [iso_post]: Completed
03_update_grub_efi.sh [iso_post]: Starting
    Creating new image: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/new.img
    Emptying and creating grub module dir: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/x86_64-efi
    Copying from grub_src_dir: /usr/lib/grub/x86_64-efi
    Emptying and creating grub module dir: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/i386-efi
    Copying from template_dir: /home/rakesh/Downloads/iso/rdp/ISO/extract/iso-directory-structure/boot/grub/x86_64-efi
    Creating EFI image: bootx64.efi
    Creating grubx64.efi
    Creating EFI image: bootia32.efi
    grub-mkimage: error: cannot open `/usr/lib/grub/i386-efi/moddep.lst': No such file or directory.
    Space remaining in new image: 2394112 bytes
    Cleaning up mounts and loop devices
03_update_grub_efi.sh [iso_post]: Completed
(run_remaster_commands): Completed
(update_squashfs): Creating squashfs --> filesystem.squashfs
(update_squashfs): Completed
(update_iso): Creating ISO ... /home/rakesh/Downloads/iso/rdp/ISO/out/modified.iso
(update_iso): Creating MBR- and EFI-compatible ISO
(update_iso): Completed

--------------------------------------------------------------------------
Source ISO=/home/rakesh/Downloads/iso/rdp/ISO/in/source.iso
Output ISO=/home/rakesh/Downloads/iso/rdp/ISO/out/modified.iso
VolID=Ubuntu 18.04.4 LTS amd64

--------------------------------------------------------------------------

Start: Wed Mar  4 16:23:47 IST 2020
Ended: Wed Mar  4 16:46:47 IST 2020
```

**Comment by sundarnagarajan** [comment]
**First**:
An elementary question:

Are you sure the ```modified.iso``` file is not present in the indicated path: ```/home/rakesh/Downloads/iso/rdp/ISO/out/modified.iso``` ?

**Second**: I notice the following error output in ```03_update_grub_efi.sh```:
**grub-mkimage: error: cannot open `/usr/lib/grub/i386-efi/moddep.lst': No such file or directory.**

Can you check the following on your **HOST** machine:
- Is this file present: ```/usr/lib/grub/i386-efi/moddep.lst``` ?
- If the file is **NOT** present, is the package ```grub-efi-ia32-bin``` installed on the **HOST** machine?  You can check this with: ``` dpkg -l | grep '^ii' | fgrep -qx grub-efi-ia32-bin```

If the output ISO ```modified.iso``` is indeed missing, this is a bug.

**Comment by rakeshtembhurne** [comment]
I am 100% sure that the `modified.iso` or any other file is not present in `/home/rakesh/Downloads/iso/rdp/ISO/out/` folder. I ran the script twice and got the same result both the time. 

No, the file `moddep.lst` is not present at the given location. Infact there is no `i368-efi` folder in `/usr/lib/grub` folder. Here is the output of the `tree -d /usr/lib/grub/` command:
```
/usr/lib/grub/
├── x86_64-efi
└── x86_64-efi-signed
```

The command `dpkg -l | grep '^ii' | fgrep -qx grub-efi-ia32-bin` prints nothing, so `grub-efi-ia32-bin` is not installed.

**Comment by rakeshtembhurne** [comment]
I tested further, and I found that the problem is with `-qx` parameters in `fgrep` command.

```
➜  dpkg -l | grep '^ii' | fgrep -qx grub-efi-ia32-bin 
➜  dpkg -l | grep '^ii' | fgrep -qx grub-efi-amd64-bin
➜  dpkg -l | grep '^ii' | fgrep grub-efi-ia32-bin 
➜  dpkg -l | grep '^ii' | fgrep grub-efi-amd64-bin
ii  grub-efi-amd64-bin                                          2.02-2ubuntu8.14                                         amd64        GRand Unified Bootloader, version 2 (EFI-AMD64 binaries)

fgrep --version
grep (GNU grep) 3.1

grep --version
grep (GNU grep) 3.1

```

**Comment by sundarnagarajan** [comment]
My mistake on the ```fgrep -qx``` command line. The correct command should be:
```dpkg -l | grep '^ii' | sed -e '1,5d' | awk '{print $2}' | fgrep -qx grub-efi-ia32-bin```

The package ``` grub-efi-ia32-bin``` is **REQUIRED on the HOST** and the script where it is used - ```03_update_grub_efi.sh``` - is run **OUTSIDE the chroot**.

Surprisingly, ```make_rdp_iso.sh``` **DOES** check for required packages and reports that ``` grub-efi-ia32-bin``` is installed !

```
All required packages are already installed
Required packages:
    grub-efi-ia32-bin grub-efi-amd64-bin grub-pc-bin grub2-common
    grub-common util-linux parted gdisk mount xorriso genisoimage
    squashfs-tools rsync git build-essential kernel-package fakeroot
    libncurses5-dev libssl-dev ccache libfile-fcntllock-perl curl
```

My conclusion is that somehow your installation of package ```grub-efi-ia32-bin``` on  your **HOST** machine is bad.

In particular, (at least) the following directory should be present on your **HOST** machine:
```/usr/lib/grub/i386-efi/```
and (at least) the following file should be present on your **HOST** machine within that directory:
```/usr/lib/grub/i386-efi/moddep.lst```
You also require many of the '.mod' files within the directory.

From your previous reply, this directory and this file are somehow missing.

You can check the integrity of all the files in package ```grub-efi-ia32-bin``` using the command:
```( cd / ; md5sum -c /var/lib/dpkg/info/grub-efi-ia32-bin.md5sums 1>/dev/null ; [ $? -eq 0 ] && echo "OK" || echo "Failed" )```

I have put in more robust checks to catch such situations in the future.

You can pull the updated ```make_rdp_iso.sh``` and run it - it will pull the new versions with better checks and it will fail early and then you will need to execute:
```sudo apt-get install --reinstall grub-efi-ia32-bin```

Once ```make_rdp_iso.sh``` no longer flags an early error, your ISO should be created.

**Comment by sundarnagarajan** [comment]
Any progress?

**Comment by rakeshtembhurne** [comment]
I pulled the latest `make_rdp_iso.sh` file and re-run the script. It worked well. I was able to install Ubuntu 18.04.4 with it. It also showed missing packages info and gave a command to install them manually.  I confirm that the issue has been solved and the ticket can be closed from my side. 

@sundarnagarajan Thank you so much for creating this repo, making it available to us and your hard work!
